yktracec/src/jitmodbuilder.cc

```diff
@@ -1732,6 +1738,10 @@ class JITModBuilder {
             }
             CallStack.pushFrame(StackFrame::CreateForeignFrame());
             break;
+          } else if (CF->getName() == PromoteRecFnName) {
+            // A variable is being promoted to a constant.
```

Comment:
I don't think this is quite correct: the result of an expression is being promoted to a constant. I can do `x = yk_promote(1)` (which is silly, I admit) or `x = yk_promote(1 + y)`, neither of which is a variable.

Reply:
Yes, I'd never thought of it like this.

So when we have `y = yk_promote(x)` we actually promote the (expression? lval?) `y` with the value of computing the *expression* `x`...

Do we want to allow things like:

```
return yk_promote(x + 1 / f());
```

?

We can reject that sort of thing if we don't want it, but at the same time, it will probably work...

If we want to allow it, I think we need to add some tests.

Reply:
I guess my question is: did you implement promoting of variables or values? I can see arguments for either of them, but I'm not sure which the PR actually does!

Reply:
I intended to implement the former, but I think I've unwittingly implemented the latter. I'll experiment tomorrow.

Reply:
Is the difference important? I'm not sure it is, at least not at the moment.

Reply:
I don't think it matters! The latter is more flexible for sure.

Reply:
Yes, but let me add tests to ensure it does actually work :)

Reply:
I assumed `yk_promote(x + y)` will just just promote whatever LLVM turns that expression into. E.g. in LLVM it would look like this:

```
$3 = add $1, $2
promote($3)
```

So it should just work™?

Reply:
Spoiler: it does :)

Reply:
Test showing it works in 27b51e994f1bec0c07211ec4e7f6edecb990f879

Reply:
At the llvm level, yes, you pass an SSA variable to `yk_promote` and receive back a distinct SSA variable which you should then use henceforth.

Reply:
OK, so imagine I have this C code:

```
x = 1 + 2 + 3;
y = promote(x);
f(y);
f(x);
```

When I call `f(x)` is that with a promoted value (in the sense that I've promoted `x` indirectly) or is only `f(y)` with a promoted value?

Reply:
Only `f(y)` I reckon.

Do you agree @ptersilie?

[In your example `promote()` probably slows the code down, because `promote` is opaque and then
 the compiler can't assume `x==y` and that `y` is statically constant :)]

Reply:
If this is the case (and I suspect it is!) can we add a test that checks so?

Reply:
Tested in 04790bbf4de1566785ba3803c7049a075edf5872

Reply:
@vext01 Perfect -- thanks!